### DISCUSSION TITLE
Block editor with AI autocompletions

### DISCUSSION THREAD

**Discussion by danielduckworth** [directus/directus#19341] [Feature Requests]
### Summary

With the [AI Hackathon announcement](https://docs.directus.io/blog/announcing-directus-ai-hackathon.html), and the introduction of [Websockets in 10.3.0](https://github.com/directus/directus/releases/tag/v10.3.0), streaming AI autocompletions for the block editor (like theNotion block editor) is a natural fit. 

### Basic Example

Allow the user to use a key to trigger AI autocompletions in the block editor.

Vercel have a [great starter UI called Novel](https://novel.sh/) which is built with TipTap and the [Vercel AI SDK](https://sdk.vercel.ai/docs) which could serve as a reference.

### Motivation

Products with AI assisted content creation could become a significant deciding factor for users. Platforms that fail to find ways to integrate the technology will likely loose market share.

### Detailed Design

Steven Tey's repo is a great implementation: https://github.com/steven-tey/novel.

Of course, Vercel all about React and Next.js, but their AI SDK supports Vue (not sure how well).

Vercel AI SDK: [https://github.com/vercel-labs/ai](https://github.com/vercel-labs/ai)

Langchain is a framework for developing apps with large language models and is not tied to any one AI technology provider.

Langchain: https://github.com/langchain-ai/langchain

Langchain JS: https://github.com/hwchase17/langchainjs

This is a [Simple example of Vue and Langchain.js](https://pr.new/https://github.com/danielduckworth/langchain-vue3-app) (set VITE_OPENAI_API_KEY env variable before running vite dev).

### Must Have:
- ENV variable to configure OpenAI API Key.
- Custom prompt field for configuring the component. This enables pre-designed prompts and "hidden prompts" that are not shown the end user, but influence the generated ouputs from the model. See OpenAI's [completions guide](https://platform.openai.com/docs/guides/gpt/chat-completions-vs-completions) and [prompt guide](https://help.openai.com/en/articles/6654000-best-practices-for-prompt-engineering-with-openai-api).
- Hotkey activated completions when using the block editor
- Slash command menu item to activate completions

### Should Have:
- The slash menu in the editor should have a field to enter custom prompt by the user.

### Could Have (features):
- Insert AI block. [This is a Notion thing](https://www.notion.so/help/creating-and-editing-with-ai#ai-content-blocks). A summarizer block can be added as a node in the block editor by one user so that when another user creates interacts with the item, they only need to click the block to generate content.

### Won't Have:
- [BabyAGI](https://agi.icils.io/)?

### Drawbacks

Other use cases like verctorized document emeddings and conversational UIs about the content stored in Directus might be more advantageos to the Directs demographic.

### Alternatives

Not sure.

### Adoption Strategy

It would be an extension to the block editor and prompted via the AI Hackathon.

**Comment by pavelstianko** [comment]
It would be awesome to have Novel in Directus!
